Lib/zipfile/__init__.py

```diff
@@ -1413,10 +1413,12 @@ def __init__(self, file, mode="r", compression=ZIP_STORED, allowZip64=True,
                 self._didModify = True
                 try:
                     self.start_dir = self.fp.tell()
+                    self._data_offset = self.start_dir
                 except (AttributeError, OSError):
                     self.fp = _Tellable(self.fp)
                     self.start_dir = 0
                     self._seekable = False
+                    self._data_offset = None
```

Comment:
It may not be set if `mode == 'a'` and if `BadZipFile` is raised. In this case, we should still have a `None` according to the docs (otherwise an AttributeError will be raised).

Reply:
I believe that in this case, since mode 'a' calls `tell()` unconditionally, we can set the `data_offset` to the result of that, right?

Reply:
Though honestly the defensive thing to do here is to initialize it to None early on then change it later when possible.

Reply:
> Though honestly the defensive thing to do here is to initialize it to None early on then change it later when possible.

Yes, that's what I thought.